📡 📉 Adding Tensorboard Tracker

sbonner0 opened the pull request:
This PR closes #383 

Adding tensorboard as a results tracker. 

What you need to test:

1. `pip install tensorboard `
2. start the tensorboard process with the default logging path:` tensorboard --logdir=.data/pykeen/logs/tensorboard/`

Issues:
- I am not happy with the current naming scheme for the experiment log dir.
- The params are current saved as text within the tb log. There is a `add_hparams` function, but I don't think it is quite what is needed here. 
- It would be great to distinguish between train and eval parameters when logging. It might also be worth considering if all the metrics actually need to be logged to tensorboard.  
- Tensorboard has great support for visualizing embeddings using the `add_embedding` function - would be great if the final embeddings (or a subset of them more realistically) could be added. 

Tasks:

- [x] add documentation and examples 
- [x] update setup.cfg with tensorboard deps

sbonner0 committed:
added tensorboard tracker file and into init

sbonner0 committed:
start of docs

cthoyt committed:
Add tutorial to index

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,71 @@
+            The additional run details which are presented as tags to be logged
+        """
+        from torch.utils.tensorboard import SummaryWriter as _SummaryWriter
+        self.SummaryWriter = _SummaryWriter
```
Probably better to store with a more variable-like name

```suggestion
        self.summary_writer_cls = _SummaryWriter
```

cthoyt replied:
You can also add an instance variable annotation in the class body after the class docstring `self.summary_writer_cls: Type['torch.utils.tensorboard.SummaryWriter']` in combination with an import inside `typing.TYPE_CHECKING`. If you're not familiar with this, I will make a commit to add it for you - it'll be quite obvious once you see it but a bit tricky to come up with yourself

sbonner0 replied:
Thanks Charlie! Would be great if you could make the commit so I can see how it is done.

cthoyt replied:
See 63eb1a4

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,71 @@
+        if experiment_path is None:
+            if experiment_name is None:
+                experiment_name = datetime.datetime.now().isoformat()
+            path = PYKEEN_LOGS / f"tensorboard/{experiment_name}"
```
That forward slash in the path part will be an issue on windows computers. Try one of the following instead:

```suggestion
            path = PYKEEN_LOGS / "tensorboard" / experiment_name
```
OR
```suggestion
            path = PYKEEN_LOGS.joinpath("tensorboard", experiment_name)
```

sbonner0 committed:
Update src/pykeen/trackers/tensorboard.py

Using joinpath instead of /

Co-authored-by: Charles Tapley Hoyt <cthoyt@gmail.com>

cthoyt committed:
Add type annotation in class for instance variable SummaryWriter

This commit does the following:
1. Imports TYPE_CHECKING from typing. This is always false during Python runtime, but set to true when a type checker like MyPy is being used to statically analyze the code
2. Uses a conditional to wrap importing the tensorboard stuff at the top. this means that during actual runtime, this is never imported
3. Make a type annotation for instance variables in the class. Specifically, this is for the SummaryWriter variable. Note that the thing that's stored in this variable is the class itself, so to refer to the class and not it being an instance of the class, we wrap the annotation with Type[]. Further, note that the annotation is done as a string. This makes it a lazy type annotation, which is good because it won't explode because it points to a package that will not be imported.

Note: this is independent from my other comment, where I said this isn't really a good name for an instance variable.

sbonner0 committed:
Update src/pykeen/trackers/tensorboard.py

better name for the summary writer

Co-authored-by: Charles Tapley Hoyt <cthoyt@gmail.com>

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,78 @@
+            path = PYKEEN_LOGS.joinpath("tensorboard", experiment_name)
+        elif isinstance(experiment_path, str):
+            path = pathlib.Path(experiment_path)
+        self.path = path
```
what should happen if the path already exists?

sbonner0 replied:
It's a very good point and something I am not quite sure how to handle. If the user runs two jobs passing the same directory for tb, you can get some interesting results as seen below:
![Screenshot from 2021-05-06 20-14-16](https://user-images.githubusercontent.com/10208489/117353218-63a93a00-ae9f-11eb-8efd-719a9285a69f.png)
Of course this is desirable behavior if you are resuming training. 

Ideally, the structure should be: `root_exp_dir/exp_1.... root_exp_dir/exp_n.` Then tb could be run at the root dir and allow for comparison between different experiments where some parameters are being changed for example.

Any thoughts on how this should be handled?

sbonner0 replied:
Were there any thoughts from anyone on how this should be handled? I am inclined towards trusting the user to set the directory correctly but very open to suggestions.

cthoyt replied:
I'd rather not over-engineer this. Just have make sure the docs make it clear that it's up to the user to take care of this

mberr commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,78 @@
+    """A tracker for TensorBoard."""
+
+    #: The class that's used to instantiate a summarywriter
+    SummaryWriter: Type['torch.utils.tensorboard.SummaryWriter']
```
should this be `summary_writer_cls`?

sbonner0 replied:
I am actually not sure, this was a change Charlie added.

mberr replied:
@cthoyt 🙂

cthoyt replied:
yes this should also be changed

cthoyt replied:
https://github.com/sbonner0/pykeen/commit/dec8030c760ae077967201d29b8e0e215f724462

sbonner0 committed:
update start_run to reflect new naming

cthoyt committed:
Update tensorboard.py

sbonner0 committed:
expand docs with log dir arg

sbonner0 committed:
more docs

mberr committed:
Update using_tensorboard.rst

indention in code example

mberr commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,78 @@
+
+        if experiment_path is None:
+            if experiment_name is None:
+                experiment_name = datetime.datetime.now().isoformat()
```
```suggestion
                experiment_name = datetime.datetime.now().isoformat().replace(":", "_")
```

Directly using [datetime.datetime.isoformat](https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat) as filename will result in a name such as `2019-05-18T15:17:08.132263` including `:`, which is a reserved character under Windows (cf. [here](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file?redirectedfrom=MSDN#file_and_directory_names)).

cthoyt replied:
We have code elsewhere in the package for making dates that are appropriate for file paths. Maybe that should be living in the `pykeen.utils` so it can be reused

sbonner0 replied:
Should I hold off on committing this then?

cthoyt replied:
Hmm now that I made two commits, I think this would just be overdoing it with functions. Just use 
```python
import time
datetime_str = time.strftime('%Y-%m-%d-%H-%M')
```

cthoyt committed:
Update utils.py

cthoyt committed:
Update ablation.py

cthoyt committed:
Merge branch 'master' into feature/tensorboard

PyKEEN-bot committed:
Trigger CI

sbonner0 committed:
fixed time string and added docs on starting tensorboard server

cthoyt committed:
Revert

cthoyt committed:
Merge branch 'master' into feature/tensorboard

cthoyt committed:
Update utils.py

cthoyt committed:
Update __init__.py

cthoyt commented on docs/source/tutorial/trackers/using_tensorboard.rst:
```diff
@@ -0,0 +1,73 @@
+It is advisable to use a unique sub-directory for each experiment to allow for easy comparison.
+
+Starting a Tensorboard Server
+-----------------
```
move the instructions for installing and running tensorboard up to the top (see MLFlow's docs as an example) so when someone like me, who's never used tensorboard, reads this, they don't accidentally run some code that wont work

cthoyt commented on docs/source/tutorial/trackers/using_tensorboard.rst:
```diff
@@ -0,0 +1,73 @@
+        result_tracker='tensorboard',
+    )
+
+It is placed in a subdirectory of :mod:`pystow` default data directory with PyKEEN called ``tensorboard``,
```
Is this a setting baked into the code, or is this dependent on how you ran tensorboard from the command line?

I don't understand what the argument given to `--logdir` in the `tensorboard` command does

cthoyt replied:
Is it the case that the python code interacts directly with the filesystem and then the frontend also interacts with the filesystem, and not that it interacts with tensorboard through an API?

sbonner0 replied:
So the python code interacts directly with the filesystem and writes the logs out as TFEvent files. These log files are then read by the `tensorboard` process to be displayed in the browser.

To be clear, the tensorboard frontend does not need to be running at the same time as the training. So the `--logdir` argument is simply telling TB where to look for the TFEvent files to display.  These logs can be copied from machine to machine as well - this is often how I use it as I train on the cluster, but want to run TB from my own machine.

cthoyt replied:
Really interesting! This is different from how the other trackers for Weights and Biases and MLflow work, so definitely worth including in the tutorial

cthoyt committed:
Remove trailing whitespace from template

cthoyt committed:
Regenerate README

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,78 @@
+        self.path = path
+
+    def start_run(self, run_name: Optional[str] = None) -> None:  # noqa: D102
+        self.writer = self.summary_writer_cls(log_dir=self.path, comment=run_name)
```
The documentation for SummarWriter says the following about the `comment` argument:
```
comment (string): Comment log_dir suffix appended to the default
              ``log_dir``. If ``log_dir`` is assigned, this argument has no effect.
```

Is this actually doing anything here?

cthoyt replied:
Also I think the writer should be assigned in the `__init__`

sbonner0 replied:
Thanks - I have corrected both now, although there is now no start_run overwritten in this class....not sure if that is normal for these trackers.

cthoyt replied:
Sorry I missed this comment before. It's okay if you don't use part of the functionality

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,78 @@
+            if experiment_name is None:
+                experiment_name = time.strftime('%Y-%m-%d-%H-%M-%S')
+            path = PYKEEN_LOGS.joinpath("tensorboard", experiment_name)
+        elif isinstance(experiment_path, str):
```
what happens if neither of these is true? `path` will be unassigned. Should it be assigned the value for `experiment_path`?

sbonner0 replied:
If the user was passing a `experiment_path` variable that wasn't a string, I wasn't sure how to handle it. I have added a condition in there now to catch that case, but it doesn't seem very elegant. Very open to suggestions!

cthoyt replied:
I reorganized this in b9f56110

sbonner0 committed:
added tensorboard to setup.cfg

sbonner0 committed:
updated the docs with install

sbonner0 committed:
remove redundant use of comment arg

sbonner0 committed:
moved writer into init

sbonner0 committed:
added else to catch edge cases on paths

sbonner0 committed:
added hpo example to docs

cthoyt committed:
Merge branch 'master' into feature/tensorboard

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Add tests

cthoyt committed:
Update installation extras table

cthoyt committed:
Minor editing and revising

Trigger CI

cthoyt committed:
Update CHANGELOG.rst

cthoyt committed:
Simplify storage of path

cthoyt committed:
Simplify storage and annotation of summary writer

cthoyt committed:
Path isn't reused so don't save it

cthoyt commented on src/pykeen/trackers/tensorboard.py:
```diff
@@ -0,0 +1,79 @@
+            The additional run details which are presented as tags to be logged
+        """
+        import torch.utils.tensorboard
+        self.tags = tags
```
Everything looks good to go on this PR except the tags. they appear to be ignored. Do you know where these should be used @sbonner0?

sbonner0 replied:
Hi @cthoyt, I think these tags are just a hangover from when I used the mlflow file as a template. I'll remove them.

sbonner0 committed:
removed unused tag arg

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Merge branch 'master' into feature/tensorboard

PyKEEN-bot committed:
Trigger CI
